kernel/printk/printk.c

```diff
@@ -3319,4 +3357,211 @@ void show_regs_print_info(const char *log_lvl)
 	       log_lvl, current, task_stack_page(current));
 }
 
+#ifdef CONFIG_LOGBUFFER
+
+/* Bootloader log control block address passed in on the kernel cmdline */
+phys_addr_t log_cb_phy_addr __initdata = 0;
+
+/* save requested log_cb_phy_addr since it's too early to process it */
+static int __init log_cb_phy_addr_setup(char *str)
+{
+	log_cb_phy_addr = memparse(str, &str);
+
+	return 0;
+}
+early_param("log_cb_phy_addr", log_cb_phy_addr_setup);
+
+static struct printk_log *log_from_idx_and_base(const char * const base, const u32 idx)
+{
+	struct printk_log *msg = (struct printk_log *)(base + idx);
+
+	if (!msg->len || msg->magic != LOGBUFF_MAGIC )
+		return (struct printk_log *)base;
+	return msg;
+}
+
+static u32 log_next_idx_from_idx(const char * const base, const u32 idx)
+{
+	struct printk_log const * const msg = (struct printk_log *)(base + idx);
+
+	if (!msg->len || msg->magic != LOGBUFF_MAGIC )
+		return 0;
+	return idx + msg->len;
+}
+
+static bool is_valid_external_cb ( const log_cb_t const * new_cb )
+{
+	// Validate the new control buffer
+	return (new_cb &&
+		new_cb->log_version == LOGBUFF_LOG_VERSION &&
+		new_cb->magic == LOGBUFF_MAGIC &&
+		new_cb->log_cb_pad_length ==
+		LOGBUFF_CB_PAD_LENGTH &&
```

Comment:
Purely stylistic, but any reason to break this and the one below on line 3401 over two lines?